### PR TITLE
[MLIR] Properly add operations to blocks during `createOrFold`

### LINKED ISSUE #68884: [MLIR] `createOrFold`-created ops has null block when accessing folder
Submitting as an issue, since this seems like a bug to me;

When an operation is created through `createOrFold`, the operation will
1. first be created
2. attempted folded
3. inserted into the block

https://github.com/llvm/llvm-project/blob/f89b7a9ee205acbd6c180cc56ea5a74f8602d269/mlir/include/mlir/IR/Builders.h#L515-L518

One may have a folder which also looks at the `mlir::Block`s where e.g. the operands of the newly created operation resides, and compare this wrt. the `mlir::Block` that the newly created operation resides in. This - to me - shouldn't violate any existing constraints (e.g. IsolatedFromAbove, local transformation, ...) since we're just looking across blocks.   

However, this may result in a folder not applying, seeing as, in step 2, when an op is created through `createOrFold`, the `block` of an op is still a `nullptr` when the fold method is run - given that the `block` pointer of an op is updated in step 3, after folding failed.

Another hypothetical issue along the lines of this is if a folder for some reason has the op block iterator as load bearing (e.g. relative position wrt. other ops); this, as well, is not assigned until after the op is inserted.

Solutions could be that , `createOrFold` pre-emptively assign the `block` pointer the op in-between step 1 and 2 (wouldn't fix the block iterator issue) or that we "fully" create an operation (e.g. insert it into the block) before applying the folder method (which then may remove it from the block again).

### PR DESCRIPTION
**Pull request by mortbopet**
Fixes #68884.

### CODE REVIEW

File: mlir/include/mlir/IR/Builders.h [hunk 1]
```diff
@@ -504,18 +504,20 @@ class OpBuilder : public Builder {
...
+    // the listener is notified.
     OperationState state(location,
                          getCheckRegisteredInfo<OpTy>(location.getContext()));
     OpTy::build(*this, state, std::forward<Args>(args)...);
     Operation *op = Operation::create(state);
+    if (block)
+      block->getOperations().insert(insertPoint, op);
 
-    // Fold the operation. If successful destroy it, otherwise insert it.
+    // Fold the operation. If successful destroy it, otherwise notify.
```
Comment (matthias-springer) [inline mlir/include/mlir/IR/Builders.h:516]: `destroy` -> `erase`

File: mlir/test/lib/Dialect/Test/TestDialect.cpp [hunk 2]
```diff
@@ -529,6 +529,10 @@ LogicalResult TestOpWithVariadicResultsAndFolder::fold(
 }
 
 OpFoldResult TestOpInPlaceFold::fold(FoldAdaptor adaptor) {
+  // Excercise the fact that an operation created with createOrFold should be
```
Comment (matthias-springer) [inline mlir/test/lib/Dialect/Test/TestDialect.cpp:532]: `Exercise`

File: mlir/test/lib/Dialect/Test/TestDialect.cpp [hunk 3]
```diff
@@ -529,6 +529,10 @@ LogicalResult TestOpWithVariadicResultsAndFolder::fold(
...
+  // allowed to enable its block.
```
Comment (matthias-springer) [inline mlir/test/lib/Dialect/Test/TestDialect.cpp:533]: What does it mean to "enable a block"?

File: mlir/test/lib/Dialect/Test/TestDialect.cpp [hunk 4]
```diff
@@ -529,6 +529,10 @@ LogicalResult TestOpWithVariadicResultsAndFolder::fold(
...
+  assert(getBlock() && "expected block to be assigned");
```
Comment (matthias-springer) [inline mlir/test/lib/Dialect/Test/TestDialect.cpp:534]: Maybe rephrase as `expected that operation is not unlinked`

File: mlir/test/lib/Dialect/Test/TestDialect.cpp (hunk 3 above)
Comment (mortbopet) [reply mlir/test/lib/Dialect/Test/TestDialect.cpp:533]: Typo, meant 'access' its block.

**Review by joker-eph — APPROVED** [review]
